File: legacy/calcD.m
function [d,costheta,LWoverlap] = calcD(x,y,z,We,g)
N=length(x);

d=zeros(1,N*(N-1)/2);
costheta=zeros(1,N*(N-1)/2);

d0=zeros(1,N*(N-1)/2);

Linewidth_d=zeros(1,N*(N-1)/2);
LWoverlap=zeros(1,N*(N-1)/2);


[j,i]=find(tril(ones(N),-1));  %i is row, j is column
ind=[i,j]';
%dxy(:)=sqrt( (x(ind(1,:)) -x(ind(2,:)) ).^2 +(y(ind(1,:)) -y(ind(2,:)) ).^2 );
d(:)=sqrt( (x(ind(1,:)) -x(ind(2,:)) ).^2 +(y(ind(1,:)) -y(ind(2,:)) ).^2 +(z(ind(1,:)) -z(ind(2,:)) ).^2);
d=d';


Linewidth_d(:)=We(ind(1,:))-We(ind(2,:));
LWoverlap= 4*g^2./(Linewidth_d.^2+4*g^2);


costheta(:)= (y(ind(1,:)) -y(ind(2,:)) )./d';


costheta=costheta';
end

File: src/simulation/utils/calcD.py
import torch
import numpy as np

def calc_D (x, y, z, We, g):
    N = len(x)
    # M = int(N*(N-1)/2)
    # lw = torch.zeros(1, M)    
    j, i = np.tril_indices(N, -1, N)
    # print(j)
    # print(i)
    d = np.sqrt(np.power((x[i]-x[j]),2) + np.power((y[i]-y[j]),2) + np.power((z[i]-z[j]),2))
    lw_overlap = 4*g**2/(np.power(We[i] - We[j],2)+4*g**2)
    costheta = np.divide(np.transpose(y[i]-y[j]), np.transpose(d))

    return np.transpose(d), costheta, np.transpose(lw_overlap)